File: java/core/src/org/daisy/urakawa/property/PropertyFactory.java
package org.daisy.urakawa.property;

import org.daisy.urakawa.GenericWithPresentationFactory;
import org.daisy.urakawa.Presentation;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.property.channel.ChannelsProperty;
import org.daisy.urakawa.property.xml.XmlProperty;

/**
 * Extension of the generic factory to handle one or more specific types derived
 * from the base specified class, in order to provide convenience create()
 * methods.
 * 
 * @xhas - - 1 org.daisy.urakawa.Presentation
 * @depend - Create - org.daisy.urakawa.property.Property
 * @depend - Create - org.daisy.urakawa.property.xml.XmlProperty
 * @depend - Create - org.daisy.urakawa.property.channel.ChannelsProperty
 */
public final class PropertyFactory extends GenericWithPresentationFactory<Property>
{
    /**
     * @param pres
     * @throws MethodParameterIsNullException
     */
    public PropertyFactory(Presentation pres)
            throws MethodParameterIsNullException
    {
        super(pres);
    }
    /**
     * @return
     */
    public Property createProperty()
    {
        try
        {
            return create(Property.class);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }

    /**
     * @return
     */
    public ChannelsProperty createChannelsProperty()
    {
        try
        {
            return create(ChannelsProperty.class);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }

    /**
     * @return
     */
    public XmlProperty createXmlProperty()
    {
        try
        {
            return create(XmlProperty.class);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }
}

File: csharp/core/property/PropertyFactory.cs
using urakawa.property.channel;

namespace urakawa.property
{
    /// <summary>
    /// Factory for creating <see cref="Property"/>s and derived types
    /// </summary>
    public sealed class PropertyFactory : GenericWithPresentationFactory<Property>
    {

        /// <summary>
        /// Creates a <see cref="ChannelsProperty"/>
        /// </summary>
        /// <returns>The created <see cref="ChannelsProperty"/></returns>
        public ChannelsProperty CreateChannelsProperty()
        {
            return Create<ChannelsProperty>();
        }

        /// <summary>
        /// Creates an <see cref="urakawa.property.xml.XmlProperty"/> instance
        /// </summary>
        /// <returns>The created instance</returns>
        public xml.XmlProperty CreateXmlProperty()
        {
            return Create<xml.XmlProperty>();
        }
    }
}